### PR TITLE
32 registration page

### PR DESCRIPTION
**Pull request by KarnaushenkoVO**
## TRELLO

* [Main Trello ticket](https://trello.com/c/pJh1wFkQ/32-create-registration-page)

## Code reviewers

- [ ] @Ashiato
- [ ] @black-art
- [x] @hrytskivr
- [ ] @KarnaushenkoVO
- [ ] @PanLeon
- [ ] @spycheast
- [x] @yuriiZherebukh

### Second Level Review

- [ ] @byshovets
- [ ] @lhalam

## Summary of issue

Added user activation via email
Added daemon for cleaning not activated users
Configured webpack with file-loader

## Testing approach

unit tests

## CHECK LIST
- [ ]  СI passed
- [ ]  Сode coverage >=90%
- [ ]  PR is reviewed manually again (to make sure you have 100% ready code)
- [ ]  All reviewers agreed to merge the PR
- [ ]  I've checked new feature as logged in and logged out user if needed
- [ ]  PR meets all conventions

### CODE REVIEW

File: myTrip/registration/models.py [hunk 1]
```diff
@@ -183,3 +220,45 @@ def get_by_facebook_id(facebook_id):
...
+    create_at = models.DateTimeField()
+
+    @staticmethod
+    def create(user, hash):
+        """
+        Creates and saves a User with the given email and password.
+        Args:
+            user (Object<CustomUser>): new user.
+        Returns:
+           hash (str): for appropriate user.
```
Comment (lhalam) [inline myTrip/registration/models.py]: update

File: myTrip/registration/models.py [hunk 2]
```diff
@@ -183,3 +220,45 @@ def get_by_facebook_id(facebook_id):
...
+    def get_user_by_hash(hash):
+        """
+        Gives user by hash
+        Args:
+            user (Object<CustomUser>): new user.
+        Returns:
+           hash (str): for appropriate user.
+        """
+
+        hash_user = HashUser.objects.get(hash=hash)
```
Comment (lhalam) [inline myTrip/registration/models.py]: try

File: myTrip/registration/views.py [hunk 3]
```diff
@@ -18,6 +20,15 @@
 FACEBOOK_AUTH_URL = ('https://www.facebook.com/v2.10/dialog/oauth?'
                      'client_id={client_id}&redirect_uri={redirect_uri}')
 
+LOGIN_URL = 'http:triptrck.com/login/'
```
Comment (lhalam) [inline myTrip/registration/views.py]: fix

File: myTrip/registration/views.py [hunk 4]
```diff
@@ -45,8 +56,14 @@ def register(request):
         if CustomUser.get_by_email(email):
             return response_400_already_registered
 
-        CustomUser.create(email=email, password=password,
-                          first_name=first_name, last_name=last_name)
+        user = CustomUser.create(email=email, password=password,
```
Comment (lhalam) [inline myTrip/registration/views.py:59]: if

File: myTrip/tests/unittests/registration/test_views.py [hunk 5]
```diff
@@ -143,11 +151,18 @@ def test_facebook_auth(self):
...
 
         request = self.client.get(FACEBOOK_LOGIN_URL)
         self.assertEqual(request.status_code, 302)
+
+    def test_activation(self):
+        """
+        Test activation via email
+        """
+        request = self.client.get(ACTIVATION_URL)
+        self.assertEqual(request.status_code, 200)
```
Comment (lhalam) [inline myTrip/tests/unittests/registration/test_views.py]: endline

File: myTrip/registration/views.py [hunk 6]
```diff
@@ -116,15 +133,29 @@ def facebook_login(request):
         redirect_uri=FACEBOOK_REDIRECT_URL,
         client_secret=CLIENT_SECRET,
         code=code)
-    token_request = urllib.request.urlopen(link)
+    token_request = urlopen(link)
     token = loads(token_request.read().decode('utf-8')).get('access_token')
-    user_request = urllib.request.urlopen(FACEBOOK_USER_URL.format(token=token))
+    user_request = urlopen(FACEBOOK_USER_URL.format(token=token))
     user_data = loads(user_request.read().decode('utf-8'))
     facebook_id = str(user_data['id'])
```
Comment (lhalam) [inline myTrip/registration/views.py]: get

File: myTrip/registration/views.py [hunk 7]
```diff
@@ -116,15 +133,29 @@ def facebook_login(request):
...
+
+def activation(request):
+    """Activation view activate user"""
+
+    hash = request.GET.get('hash')
+    if not hash:
+        return HttpResponse(status=400)
+    user = HashUser.get_user_by_hash(hash)
+    if not user:
+        return HttpResponse(status=404)
```
Comment (lhalam) [inline myTrip/registration/views.py]: statu=400

File: myTrip/registration/models.py [hunk 8]
```diff
@@ -183,3 +220,48 @@ def get_by_facebook_id(facebook_id):
...
+class HashUser(models.Model):
+    """
+    Table for finding active users
+    :argument hash: str - activation hash for every user
+    :argument user: Object<CustomUser>: - foreign key to CustomUser model
+    """
+
+    hash = models.CharField(max_length=500, blank=False)
+    user = models.OneToOneField(CustomUser, blank=False)
+    create_at = models.DateTimeField()
```
Comment (lhalam) [inline myTrip/registration/models.py]: avto

File: myTrip/registration/models.py [hunk 9]
```diff
@@ -183,3 +220,48 @@ def get_by_facebook_id(facebook_id):
...
+            user (Object<CustomUser>): new user.
+            hash (str): for appropriate user.
+        Returns:
+           hash_user (Object<HashUser>): new hash-user .
+        """
+
+        hash_user = HashUser()
+        hash_user.user = user
+        hash_user.hash = hash
+        hash_user.create_at = datetime.datetime.now()
```
Comment (lhalam) [inline myTrip/registration/models.py]: delete

**Review by hrytskivr — APPROVED** [review]
погоджено

**Pull request merged**